File: firefly-example/src/main/java/com/firefly/example/reactive/coffee/store/dao/ProductDAO.java
package com.firefly.example.reactive.coffee.store.dao;

import com.firefly.example.reactive.coffee.store.model.Product;
import com.firefly.example.reactive.coffee.store.vo.Page;
import com.firefly.example.reactive.coffee.store.vo.ProductQuery;
import com.firefly.reactive.adapter.db.ReactiveSQLConnection;
import reactor.core.publisher.Mono;

import java.util.List;

/**
 * @author Pengtao Qiu
 */
public interface ProductDAO {

    Mono<Page<Product>> list(ProductQuery query);

    Mono<Product> get(Long id);

    Mono<List<Product>> list(List<Long> idList, ReactiveSQLConnection connection);

    Mono<Long> insert(Product product);

    Mono<Integer> update(Product product);
}

File: firefly-example/src/main/kotlin/com/firefly/example/kotlin/coffee/store/dao/ProductDAO.kt
package com.firefly.example.kotlin.coffee.store.dao

import com.firefly.example.kotlin.coffee.store.model.Product
import com.firefly.example.kotlin.coffee.store.vo.Page
import com.firefly.example.kotlin.coffee.store.vo.ProductQuery

/**
 * @author Pengtao Qiu
 */
interface ProductDAO {

    suspend fun list(query: ProductQuery): Page<Product>

    suspend fun get(id: Long): Product

    suspend fun list(idList: List<Long>): List<Product>

    suspend fun insert(product: Product): Long

    suspend fun update(product: Product): Int

}